File: py/test/selenium/webdriver/common/element_attribute_tests.py
# Licensed to the Software Freedom Conservancy (SFC) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The SFC licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.

import pytest

from selenium.common.exceptions import WebDriverException
from selenium.webdriver.common.by import By


def test_should_return_null_when_getting_the_value_of_an_attribute_that_is_not_listed(driver, pages):
    pages.load("simpleTest.html")
    head = driver.find_element(By.XPATH, "/html")
    attribute = head.get_attribute("cheese")
    assert attribute is None


def test_should_return_null_when_getting_src_attribute_of_invalid_img_tag(driver, pages):
    pages.load("simpleTest.html")
    img = driver.find_element(By.ID, "invalidImgTag")
    img_attr = img.get_attribute("src")
    assert img_attr is None


def test_should_return_an_absolute_url_when_getting_src_attribute_of_avalid_img_tag(driver, pages):
    pages.load("simpleTest.html")
    img = driver.find_element(By.ID, "validImgTag")
    img_attr = img.get_attribute("src")
    assert "icon.gif" in img_attr


def test_should_return_an_absolute_url_when_getting_href_attribute_of_avalid_anchor_tag(driver, pages):
    pages.load("simpleTest.html")
    img = driver.find_element(By.ID, "validAnchorTag")
    img_attr = img.get_attribute("href")
    assert "icon.gif" in img_attr


def test_should_return_empty_attribute_values_when_present_and_the_value_is_actually_empty(driver, pages):
    pages.load("simpleTest.html")
    body = driver.find_element(By.XPATH, "//body")
    assert "" == body.get_attribute("style")


def test_should_return_the_value_of_the_disabled_attribute_as_false_if_not_set(driver, pages):
    pages.load("formPage.html")
    inputElement = driver.find_element(By.XPATH, "//input[@id='working']")
    assert inputElement.get_attribute("disabled") is None
    assert inputElement.is_enabled()

    pElement = driver.find_element(By.ID, "peas")
    assert pElement.get_attribute("disabled") is None
    assert pElement.is_enabled()


def test_should_return_the_value_of_the_index_attribute_even_if_it_is_missing(driver, pages):
    pages.load("formPage.html")
    multiSelect = driver.find_element(By.ID, "multi")
    options = multiSelect.find_elements(By.TAG_NAME, "option")
    assert "1" == options[1].get_attribute("index")


def test_should_indicate_the_elements_that_are_disabled_are_not_is_enabled(driver, pages):
    pages.load("formPage.html")
    inputElement = driver.find_element(By.XPATH, "//input[@id='notWorking']")
    assert not inputElement.is_enabled()

    inputElement = driver.find_element(By.XPATH, "//input[@id='working']")
    assert inputElement.is_enabled()


def test_elements_should_be_disabled_if_they_are_disabled_using_random_disabled_strings(driver, pages):
    pages.load("formPage.html")
    disabledTextElement1 = driver.find_element(By.ID, "disabledTextElement1")
    assert not disabledTextElement1.is_enabled()

    disabledTextElement2 = driver.find_element(By.ID, "disabledTextElement2")
    assert not disabledTextElement2.is_enabled()

    disabledSubmitElement = driver.find_element(By.ID, "disabledSubmitElement")
    assert not disabledSubmitElement.is_enabled()


def test_should_indicate_when_atext_area_is_disabled(driver, pages):
    pages.load("formPage.html")
    textArea = driver.find_element(By.XPATH, "//textarea[@id='notWorkingArea']")
    assert not textArea.is_enabled()


@pytest.mark.xfail_safari
def test_should_throw_exception_if_sending_keys_to_element_disabled_using_random_disabled_strings(driver, pages):
    pages.load("formPage.html")
    disabledTextElement1 = driver.find_element(By.ID, "disabledTextElement1")
    with pytest.raises(WebDriverException):
        disabledTextElement1.send_keys("foo")
    assert "" == disabledTextElement1.text

    disabledTextElement2 = driver.find_element(By.ID, "disabledTextElement2")
    with pytest.raises(WebDriverException):
        disabledTextElement2.send_keys("bar")
    assert "" == disabledTextElement2.text


def test_should_indicate_when_aselect_is_disabled(driver, pages):
    pages.load("formPage.html")
    enabled = driver.find_element(By.NAME, "selectomatic")
    disabled = driver.find_element(By.NAME, "no-select")

    assert enabled.is_enabled()
    assert not disabled.is_enabled()


def test_should_return_the_value_of_checked_for_acheckbox_even_if_it_lacks_that_attribute(driver, pages):
    pages.load("formPage.html")
    checkbox = driver.find_element(By.XPATH, "//input[@id='checky']")
    assert checkbox.get_attribute("checked") is None
    checkbox.click()
    assert "true" == checkbox.get_attribute("checked")


def test_should_return_the_value_of_selected_for_radio_buttons_even_if_they_lack_that_attribute(driver, pages):
    pages.load("formPage.html")
    neverSelected = driver.find_element(By.ID, "cheese")
    initiallyNotSelected = driver.find_element(By.ID, "peas")
    initiallySelected = driver.find_element(By.ID, "cheese_and_peas")

    assert neverSelected.get_attribute("checked") is None
    assert initiallyNotSelected.get_attribute("checked") is None
    assert "true" == initiallySelected.get_attribute("checked")

    initiallyNotSelected.click()
    assert neverSelected.get_attribute("selected") is None
    assert "true" == initiallyNotSelected.get_attribute("checked")
    assert initiallySelected.get_attribute("checked") is None


def test_should_return_the_value_of_selected_for_options_in_selects_even_if_they_lack_that_attribute(driver, pages):
    pages.load("formPage.html")
    selectBox = driver.find_element(By.XPATH, "//select[@name='selectomatic']")
    options = selectBox.find_elements(By.TAG_NAME, "option")
    one = options[0]
    two = options[1]
    assert one.is_selected()
    assert not two.is_selected()
    assert "true" == one.get_attribute("selected")
    assert two.get_attribute("selected") is None


def test_should_return_value_of_class_attribute_of_an_element(driver, pages):
    pages.load("xhtmlTest.html")
    heading = driver.find_element(By.XPATH, "//h1")
    classname = heading.get_attribute("class")
    assert "header" == classname

# Disabled due to issues with Frames
# def test_should_return_value_of_class_attribute_of_an_element_after_switching_iframe(driver, pages):
#    pages.load("iframes.html")
#    driver.switch_to.frame("iframe1")
#
#    wallace = driver.find_element(By.XPATH, "//div[@id='wallace']")
#    classname = wallace.get_attribute("class")
#    assert "gromit" == classname


def test_should_return_the_contents_of_atext_area_as_its_value(driver, pages):
    pages.load("formPage.html")
    value = driver.find_element(By.ID, "withText").get_attribute("value")
    assert "Example text" == value


def test_should_return_the_contents_of_atext_area_as_its_value_when_set_to_non_norminal_true(driver, pages):
    pages.load("formPage.html")
    e = driver.find_element(By.ID, "withText")
    driver.execute_script("arguments[0].value = 'tRuE'", e)
    value = e.get_attribute("value")
    assert "tRuE" == value


def test_should_treat_readonly_as_avalue(driver, pages):
    pages.load("formPage.html")
    element = driver.find_element(By.NAME, "readonly")
    readOnlyAttribute = element.get_attribute("readonly")

    textInput = driver.find_element(By.NAME, "x")
    notReadOnly = textInput.get_attribute("readonly")

    assert readOnlyAttribute != notReadOnly


def test_should_get_numeric_attribute(driver, pages):
    pages.load("formPage.html")
    element = driver.find_element(By.ID, "withText")
    assert "5" == element.get_attribute("rows")


def test_can_return_atext_approximation_of_the_style_attribute(driver, pages):
    pages.load("javascriptPage.html")
    style = driver.find_element(By.ID, "red-item").get_attribute("style")
    assert "background-color" in style.lower()


def test_should_correctly_report_value_of_colspan(driver, pages):
    pages.load("tables.html")

    th1 = driver.find_element(By.ID, "th1")
    td2 = driver.find_element(By.ID, "td2")

    assert "th1" == th1.get_attribute("id")
    assert "3" == th1.get_attribute("colspan")

    assert "td2" == td2.get_attribute("id")
    assert "2" == td2.get_attribute("colspan")


def test_can_retrieve_the_current_value_of_atext_form_field_text_input(driver, pages):
    pages.load("formPage.html")
    element = driver.find_element(By.ID, "working")
    assert "" == element.get_attribute("value")
    element.send_keys("hello world")
    assert "hello world" == element.get_attribute("value")


def test_can_retrieve_the_current_value_of_atext_form_field_email_input(driver, pages):
    pages.load("formPage.html")
    element = driver.find_element(By.ID, "email")
    assert "" == element.get_attribute("value")
    element.send_keys("hello@example.com")
    assert "hello@example.com" == element.get_attribute("value")


def test_can_retrieve_the_current_value_of_atext_form_field_text_area(driver, pages):
    pages.load("formPage.html")
    element = driver.find_element(By.ID, "emptyTextArea")
    assert "" == element.get_attribute("value")
    element.send_keys("hello world")
    assert "hello world" == element.get_attribute("value")


def test_should_return_null_for_non_present_boolean_attributes(driver, pages):
    pages.load("booleanAttributes.html")
    element1 = driver.find_element(By.ID, "working")
    assert element1.get_attribute("required") is None


@pytest.mark.xfail_ie
def test_should_return_true_for_present_boolean_attributes(driver, pages):
    pages.load("booleanAttributes.html")
    element1 = driver.find_element(By.ID, "emailRequired")
    assert "true" == element1.get_attribute("required")
    element2 = driver.find_element(By.ID, "emptyTextAreaRequired")
    assert "true" == element2.get_attribute("required")
    element3 = driver.find_element(By.ID, "inputRequired")
    assert "true" == element3.get_attribute("required")
    element4 = driver.find_element(By.ID, "textAreaRequired")
    assert "true" == element4.get_attribute("required")


@pytest.mark.xfail_chrome
@pytest.mark.xfail_firefox
@pytest.mark.xfail_safari
@pytest.mark.xfail_remote
def test_should_get_unicode_chars_from_attribute(driver, pages):
    pages.load("formPage.html")
    title = driver.find_element(By.ID, "vsearchGadget").get_attribute("title")
    assert 'Hvad s\xf8ger du?' == title


@pytest.mark.xfail_chrome
@pytest.mark.xfail_firefox
@pytest.mark.xfail_safari
@pytest.mark.xfail_remote
def test_should_get_values_and_not_miss_items(driver, pages):
    pages.load("attributes.html")
    expected = "4b273a33fbbd29013nN93dy4F1A~"
    result = driver.find_element(By.CSS_SELECTOR, "li").get_attribute("value")
    assert expected == result

File: dotnet/test/webdriver/ElementAttributeTests.cs
// <copyright file="ElementAttributeTests.cs" company="Selenium Committers">
// Licensed to the Software Freedom Conservancy (SFC) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The SFC licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.
// </copyright>

using System.Collections.ObjectModel;

namespace OpenQA.Selenium.Tests;

[TestFixture]
public class ElementAttributeTests : DriverTestFixture
{
    [Test]
    public void ShouldReturnNullWhenGettingTheValueOfAnAttributeThatIsNotListed()
    {
        Driver.Url = Urls.SimpleTestPage;
        IWebElement head = Driver.FindElement(By.XPath("/html"));
        string attribute = head.GetAttribute("cheese");
        Assert.That(attribute, Is.Null);
    }

    [Test]
    public void ShouldReturnNullWhenGettingSrcAttributeOfInvalidImgTag()
    {
        Driver.Url = Urls.SimpleTestPage;
        IWebElement img = Driver.FindElement(By.Id("invalidImgTag"));
        string attribute = img.GetAttribute("src");
        Assert.That(attribute, Is.Null);
    }

    [Test]
    public void ShouldReturnAnAbsoluteUrlWhenGettingSrcAttributeOfAValidImgTag()
    {
        Driver.Url = Urls.SimpleTestPage;
        IWebElement img = Driver.FindElement(By.Id("validImgTag"));
        string attribute = img.GetAttribute("src");
        Assert.That(attribute, Is.EqualTo(Urls.WhereIs("icon.gif")));
    }

    [Test]
    public void ShouldReturnAnAbsoluteUrlWhenGettingHrefAttributeOfAValidAnchorTag()
    {
        Driver.Url = Urls.SimpleTestPage;
        IWebElement img = Driver.FindElement(By.Id("validAnchorTag"));
        string attribute = img.GetAttribute("href");
        Assert.That(attribute, Is.EqualTo(Urls.WhereIs("icon.gif")));
    }

    [Test]
    public void ShouldReturnEmptyAttributeValuesWhenPresentAndTheValueIsActuallyEmpty()
    {
        Driver.Url = Urls.SimpleTestPage;
        IWebElement body = Driver.FindElement(By.XPath("//body"));
        Assert.That(body.GetAttribute("style"), Is.Empty);
    }

    [Test]
    public void ShouldReturnTheValueOfTheDisabledAttributeAsNullIfNotSet()
    {
        Driver.Url = Urls.FormsPage;
        IWebElement inputElement = Driver.FindElement(By.XPath("//input[@id='working']"));
        Assert.That(inputElement.GetAttribute("disabled"), Is.Null);
        Assert.That(inputElement.Enabled, "Element is not enabled");

        IWebElement pElement = Driver.FindElement(By.Id("peas"));
        Assert.That(inputElement.GetAttribute("disabled"), Is.Null);
        Assert.That(inputElement.Enabled, "Element is not enabled");
    }

    [Test]
    public void ShouldReturnTheValueOfTheIndexAttrbuteEvenIfItIsMissing()
    {
        Driver.Url = Urls.FormsPage;

        IWebElement multiSelect = Driver.FindElement(By.Id("multi"));
        ReadOnlyCollection<IWebElement> options = multiSelect.FindElements(By.TagName("option"));
        Assert.That(options[1].GetAttribute("index"), Is.EqualTo("1"));
    }

    [Test]
    public void ShouldIndicateTheElementsThatAreDisabledAreNotEnabled()
    {
        Driver.Url = Urls.FormsPage;
        IWebElement inputElement = Driver.FindElement(By.XPath("//input[@id='notWorking']"));
        Assert.That(inputElement.Enabled, Is.False, "Element should be disabled");

        inputElement = Driver.FindElement(By.XPath("//input[@id='working']"));
        Assert.That(inputElement.Enabled, Is.True, "Element should be enabled");
    }

    [Test]
    public void ElementsShouldBeDisabledIfTheyAreDisabledUsingRandomDisabledStrings()
    {
        Driver.Url = Urls.FormsPage;
        IWebElement disabledTextElement1 = Driver.FindElement(By.Id("disabledTextElement1"));
        Assert.That(disabledTextElement1.Enabled, Is.False, "disabledTextElement1 should be disabled");

        IWebElement disabledTextElement2 = Driver.FindElement(By.Id("disabledTextElement2"));
        Assert.That(disabledTextElement2.Enabled, Is.False, "disabledTextElement2 should be disabled");

        IWebElement disabledSubmitElement = Driver.FindElement(By.Id("disabledSubmitElement"));
        Assert.That(disabledSubmitElement.Enabled, Is.False, "disabledSubmitElement should be disabled");
    }

    [Test]
    public void ShouldThrowExceptionIfSendingKeysToElementDisabledUsingRandomDisabledStrings()
    {
        Driver.Url = Urls.FormsPage;
        IWebElement disabledTextElement1 = Driver.FindElement(By.Id("disabledTextElement1"));

        Assert.That(() =>
        {
            disabledTextElement1.SendKeys("foo");
        }, Throws.TypeOf<ElementNotInteractableException>());

        Assert.That(disabledTextElement1.Text, Is.Empty);

        IWebElement disabledTextElement2 = Driver.FindElement(By.Id("disabledTextElement2"));

        Assert.That(
            () => disabledTextElement2.SendKeys("bar"),
            Throws.TypeOf<ElementNotInteractableException>());

        Assert.That(disabledTextElement2.Text, Is.Empty);
    }

    [Test]
    public void ShouldIndicateWhenATextAreaIsDisabled()
    {
        Driver.Url = Urls.FormsPage;
        IWebElement textArea = Driver.FindElement(By.XPath("//textarea[@id='notWorkingArea']"));
        Assert.That(textArea.Enabled, Is.False);
    }

    [Test]
    public void ShouldIndicateWhenASelectIsDisabled()
    {
        Driver.Url = Urls.FormsPage;

        IWebElement enabled = Driver.FindElement(By.Name("selectomatic"));
        IWebElement disabled = Driver.FindElement(By.Name("no-select"));

        Assert.That(enabled.Enabled, Is.True, "Expected select element to be enabled");
        Assert.That(disabled.Enabled, Is.False, "Expected select element to be disabled");
    }

    [Test]
    public void ShouldReturnTheValueOfCheckedForACheckboxOnlyIfItIsChecked()
    {
        Driver.Url = Urls.FormsPage;
        IWebElement checkbox = Driver.FindElement(By.XPath("//input[@id='checky']"));
        Assert.That(checkbox.GetAttribute("checked"), Is.Null);
        checkbox.Click();
        Assert.That(checkbox.GetAttribute("checked"), Is.EqualTo("true"));
    }

    [Test]
    public void ShouldOnlyReturnTheValueOfSelectedForRadioButtonsIfItIsSet()
    {
        Driver.Url = Urls.FormsPage;
        IWebElement neverSelected = Driver.FindElement(By.Id("cheese"));
        IWebElement initiallyNotSelected = Driver.FindElement(By.Id("peas"));
        IWebElement initiallySelected = Driver.FindElement(By.Id("cheese_and_peas"));

        Assert.That(neverSelected.GetAttribute("selected"), Is.Null, "false");
        Assert.That(initiallyNotSelected.GetAttribute("selected"), Is.Null, "false");
        Assert.That(initiallySelected.GetAttribute("selected"), Is.EqualTo("true"), "true");

        initiallyNotSelected.Click();
        Assert.That(neverSelected.GetAttribute("selected"), Is.Null);
        Assert.That(initiallyNotSelected.GetAttribute("selected"), Is.EqualTo("true"));
        Assert.That(initiallySelected.GetAttribute("selected"), Is.Null);
    }

    [Test]
    public void ShouldReturnTheValueOfSelectedForOptionsOnlyIfTheyAreSelected()
    {
        Driver.Url = Urls.FormsPage;
        IWebElement selectBox = Driver.FindElement(By.XPath("//select[@name='selectomatic']"));
        ReadOnlyCollection<IWebElement> options = selectBox.FindElements(By.TagName("option"));
        IWebElement one = options[0];
        IWebElement two = options[1];
        Assert.That(one.Selected, Is.True);
        Assert.That(two.Selected, Is.False);
        Assert.That(one.GetAttribute("selected"), Is.EqualTo("true"));
        Assert.That(two.GetAttribute("selected"), Is.Null);
    }

    [Test]
    public void ShouldReturnValueOfClassAttributeOfAnElement()
    {
        Driver.Url = Urls.XhtmlTestPage;

        IWebElement heading = Driver.FindElement(By.XPath("//h1"));
        String className = heading.GetAttribute("class");

        Assert.That(className, Is.EqualTo("header"));
    }

    [Test]
    public void ShouldReturnTheContentsOfATextAreaAsItsValue()
    {
        Driver.Url = Urls.FormsPage;

        String value = Driver.FindElement(By.Id("withText")).GetAttribute("value");

        Assert.That(value, Is.EqualTo("Example text"));
    }

    [Test]
    public void ShouldReturnInnerHtml()
    {
        Driver.Url = Urls.SimpleTestPage;

        string html = Driver.FindElement(By.Id("wrappingtext")).GetAttribute("innerHTML");
        Assert.That(html, Does.Contain("<tbody>"));
    }

    [Test]
    public void ShouldTreatReadonlyAsAValue()
    {
        Driver.Url = Urls.FormsPage;

        IWebElement element = Driver.FindElement(By.Name("readonly"));
        string readOnlyAttribute = element.GetAttribute("readonly");

        Assert.That(readOnlyAttribute, Is.Not.Null);

        IWebElement textInput = Driver.FindElement(By.Name("x"));
        string notReadOnly = textInput.GetAttribute("readonly");

        Assert.That(notReadOnly, Is.Null);
    }

    [Test]
    public void ShouldReturnHiddenTextForTextContentAttribute()
    {
        Driver.Url = Urls.SimpleTestPage;

        IWebElement element = Driver.FindElement(By.Id("hiddenline"));
        string textContent = element.GetAttribute("textContent");

        Assert.That(textContent, Is.EqualTo("A hidden line of text"));
    }

    [Test]
    public void ShouldGetNumericAtribute()
    {
        Driver.Url = Urls.FormsPage;
        IWebElement element = Driver.FindElement(By.Id("withText"));
        Assert.That(element.GetAttribute("rows"), Is.EqualTo("5"));
    }

    [Test]
    public void CanReturnATextApproximationOfTheStyleAttribute()
    {
        Driver.Url = Urls.JavascriptPage;
        string style = Driver.FindElement(By.Id("red-item")).GetAttribute("style");

        Assert.That(style.ToLower(), Does.Contain("background-color"));
    }

    public void ShouldCorrectlyReportValueOfColspan()
    {
        Driver.Url = Urls.Tables;
        System.Threading.Thread.Sleep(1000);

        IWebElement th1 = Driver.FindElement(By.Id("th1"));
        IWebElement td2 = Driver.FindElement(By.Id("td2"));

        Assert.That(th1.GetAttribute("id"), Is.EqualTo("th1"), "th1 id");
        Assert.That(th1.GetAttribute("colspan"), Is.EqualTo("3"), "th1 colspan should be 3");

        Assert.That(td2.GetAttribute("id"), Is.EqualTo("td2"), "td2 id");
        Assert.That(td2.GetAttribute("colspan"), Is.EqualTo("2"), "td2 colspan should be 2");
    }

    // This is a test-case re-creating issue 900.
    [Test]
    public void ShouldReturnValueOfOnClickAttribute()
    {
        Driver.Url = Urls.JavascriptPage;

        IWebElement mouseclickDiv = Driver.FindElement(By.Id("mouseclick"));

        string onClickValue = mouseclickDiv.GetAttribute("onclick");
        string expectedOnClickValue = "displayMessage('mouse click');";
        List<string> acceptableOnClickValues = new List<string>();
        acceptableOnClickValues.Add("javascript:" + expectedOnClickValue);
        acceptableOnClickValues.Add("function anonymous()\n{\n" + expectedOnClickValue + "\n}");
        acceptableOnClickValues.Add("function onclick()\n{\n" + expectedOnClickValue + "\n}");
        Assert.That(acceptableOnClickValues, Contains.Item(onClickValue));

        IWebElement mousedownDiv = Driver.FindElement(By.Id("mousedown"));
        Assert.That(mousedownDiv.GetAttribute("onclick"), Is.Null);
    }

    [Test]
    public void GetAttributeDoesNotReturnAnObjectForSvgProperties()
    {
        if (TestUtilities.IsOldIE(Driver))
        {
            Assert.Ignore("IE8 and earlier do not support SVG");
        }

        Driver.Url = Urls.SvgPage;
        IWebElement svgElement = Driver.FindElement(By.Id("rotate"));
        Assert.That(svgElement.GetAttribute("transform"), Is.EqualTo("rotate(30)"));
    }

    [Test]
    public void CanRetrieveTheCurrentValueOfATextFormField_textInput()
    {
        Driver.Url = Urls.FormsPage;
        IWebElement element = Driver.FindElement(By.Id("working"));
        Assert.That(element.GetAttribute("value"), Is.Empty);
        element.SendKeys("hello world");
        Assert.That(element.GetAttribute("value"), Is.EqualTo("hello world"));
    }

    [Test]
    public void CanRetrieveTheCurrentValueOfATextFormField_emailInput()
    {
        Driver.Url = Urls.FormsPage;
        IWebElement element = Driver.FindElement(By.Id("email"));
        Assert.That(element.GetAttribute("value"), Is.Empty);
        element.SendKeys("hello world");
        Assert.That(element.GetAttribute("value"), Is.EqualTo("hello world"));
    }

    [Test]
    public void CanRetrieveTheCurrentValueOfATextFormField_textArea()
    {
        Driver.Url = Urls.FormsPage;
        IWebElement element = Driver.FindElement(By.Id("emptyTextArea"));
        Assert.That(element.GetAttribute("value"), Is.Empty);
        element.SendKeys("hello world");
        Assert.That(element.GetAttribute("value"), Is.EqualTo("hello world"));
    }

    [Test]
    public void ShouldReturnNullForNonPresentBooleanAttributes()
    {
        Driver.Url = Urls.BooleanAttributes;
        IWebElement element1 = Driver.FindElement(By.Id("working"));
        Assert.That(element1.GetAttribute("required"), Is.Null);
        IWebElement element2 = Driver.FindElement(By.Id("wallace"));
        Assert.That(element2.GetAttribute("nowrap"), Is.Null);
    }

    [Test]
    public void ShouldReturnTrueForPresentBooleanAttributes()
    {
        Driver.Url = Urls.BooleanAttributes;
        IWebElement element1 = Driver.FindElement(By.Id("emailRequired"));
        Assert.That(element1.GetAttribute("required"), Is.EqualTo("true"));
        IWebElement element2 = Driver.FindElement(By.Id("emptyTextAreaRequired"));
        Assert.That(element2.GetAttribute("required"), Is.EqualTo("true"));
        IWebElement element3 = Driver.FindElement(By.Id("inputRequired"));
        Assert.That(element3.GetAttribute("required"), Is.EqualTo("true"));
        IWebElement element4 = Driver.FindElement(By.Id("textAreaRequired"));
        Assert.That(element4.GetAttribute("required"), Is.EqualTo("true"));
        IWebElement element5 = Driver.FindElement(By.Id("unwrappable"));
        Assert.That(element5.GetAttribute("nowrap"), Is.EqualTo("true"));
    }

    [Test]
    public void MultipleAttributeShouldBeNullWhenNotSet()
    {
        Driver.Url = Urls.SelectPage;
        IWebElement element = Driver.FindElement(By.Id("selectWithoutMultiple"));
        Assert.That(element.GetAttribute("multiple"), Is.Null);
    }

    [Test]
    public void MultipleAttributeShouldBeTrueWhenSet()
    {
        Driver.Url = Urls.SelectPage;
        IWebElement element = Driver.FindElement(By.Id("selectWithMultipleEqualsMultiple"));
        Assert.That(element.GetAttribute("multiple"), Is.EqualTo("true"));
    }

    [Test]
    public void MultipleAttributeShouldBeTrueWhenSelectHasMultipleWithValueAsBlank()
    {
        Driver.Url = Urls.SelectPage;
        IWebElement element = Driver.FindElement(By.Id("selectWithEmptyStringMultiple"));
        Assert.That(element.GetAttribute("multiple"), Is.EqualTo("true"));
    }

    [Test]
    public void MultipleAttributeShouldBeTrueWhenSelectHasMultipleWithoutAValue()
    {
        Driver.Url = Urls.SelectPage;
        IWebElement element = Driver.FindElement(By.Id("selectWithMultipleWithoutValue"));
        Assert.That(element.GetAttribute("multiple"), Is.EqualTo("true"));
    }

    [Test]
    public void MultipleAttributeShouldBeTrueWhenSelectHasMultipleWithValueAsSomethingElse()
    {
        Driver.Url = Urls.SelectPage;
        IWebElement element = Driver.FindElement(By.Id("selectWithRandomMultipleValue"));
        Assert.That(element.GetAttribute("multiple"), Is.EqualTo("true"));
    }

    [Test]
    public void GetAttributeOfUserDefinedProperty()
    {
        Driver.Url = Urls.WhereIs("userDefinedProperty.html");
        IWebElement element = Driver.FindElement(By.Id("d"));
        Assert.That(element.GetAttribute("dynamicProperty"), Is.EqualTo("sampleValue"));
    }

    [Test]
    public void ShouldReturnValueOfClassAttributeOfAnElementAfterSwitchingIFrame()
    {
        Driver.Url = Urls.IframesPage;
        Driver.SwitchTo().Frame("iframe1");

        IWebElement wallace = Driver.FindElement(By.XPath("//div[@id='wallace']"));
        String className = wallace.GetAttribute("class");
        Assert.That(className, Is.EqualTo("gromit"));
    }
}
